### PR TITLE
ci: port the merge pipeline ahead of a repository transfer

### PR DESCRIPTION
**Pull request by ivan-pinatti**
## Summary

Rehearses the whole merge pipeline that docker-torrent-box-with-vpn built over
the last day, on this repository while it is still personal owned, so that
moving it into an organization later (required before GitHub allows a merge
queue on a personal account) tests only the transfer rather than newly
written workflow code at the same time.

- `scripts/assert-pin-only-diff.py`, adapted to this repository's real pin
  surfaces: `.tool-versions`, GitHub Actions SHA pins in
  `.github/workflows/`, pre-commit `rev:` pins, and only the lines in
  `.env.example` a `# renovate:` comment actually anchors (read live off the
  checkout, not hardcoded), which is what keeps it from waving through
  `GOCRYPTFS_VERSION` or `DOCKER_IMAGE_TAG_VERSION`, both bumped by hand per
  CLAUDE.md and neither annotated for Renovate.
- `scripts/coderabbit-review-verdict.py` and `.github/workflows/coderabbit-gate.yml`,
  publishing `Pin Only` and `Review Verified` as commit statuses, ported
  essentially unchanged.
- `.github/workflows/bot-auto-merge.yml`, ported. It has nothing to approve
  until branch protection requires a review, which is a separate step the
  owner performs; said so in its header comment.
- `.github/workflows/coderabbit-review-queue.yml`, the hourly nudge, with
  `OTHER_REQUIRED_CONTEXTS` adapted to this repository's two jobs
  (`Pre-commit`, `Tests`) instead of the sibling repository's three.
- `merge_group` triggers on `pull-request-validation.yml` and
  `coderabbit-gate.yml`.
- A new `Docker Build` job, not a required check, that builds the Dockerfile
  standalone and logs into Docker Hub when `DOCKERHUB_USERNAME` /
  `DOCKERHUB_TOKEN` exist (they do not yet), skipping the login cleanly
  otherwise. It is the instrument for testing whether those secrets survive
  a future transfer, which GitHub's own docs do not answer.
- `docs/MERGE_PIPELINE.md`, this repository's own thinner version: four
  required contexts, no integration suite, and a section on what is
  deliberately not enabled yet.

Ported reasoning is attributed to docker-torrent-box-with-vpn by name in
comments rather than reusing its issue numbers as if the same incidents had
happened here.

## Test plan

- [x] `pre-commit run --all-files` passes
- [x] `pytest -m scripts tests/test_assert_pin_only_diff.py tests/test_coderabbit_review_verdict.py` (36 tests) passes
- [ ] `Pre-commit` and `Tests` checks pass on this pull request
- [ ] CodeRabbit review addressed

## Summary by CodeRabbit

- **New Features**
  - Added automated review verification and merge handling for eligible dependency updates.
  - Added safeguards to restrict dependency changes to approved version-only updates.
  - Added Docker build validation to pull request checks, including merge-queue events.

- **Documentation**
  - Added guidance covering the pull request, review, validation, and merge pipeline.

- **Tests**
  - Added comprehensive coverage for dependency-change validation, review verification, malformed input, and edge cases.

### CODE REVIEW

**Comment by ivan-pinatti** [comment]
@coderabbitai review

File: .github/workflows/bot-auto-merge.yml [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+              if [ "$state" != "pending" ]; then
+                echo "Pin Only: $(printf '%s' "$result" | jq -r '.description') ($state)"
+                [ "$state" = "success" ]
+                exit $?
+              fi
+            fi
+            sleep 5
+          done
+          echo "::error::No settled Pin Only verdict for $HEAD_SHA after two minutes."
+          exit 1
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply .github/workflows/bot-auto-merge.yml:154]: Fixed: the loop now uses `(first // empty)` before formatting, so an absent status stays empty and polling continues instead of reading it as a settled failure.

File: .github/workflows/pull-request-validation.yml [hunk 2]
```diff
@@ -52,3 +59,59 @@ jobs:
...
+          DOCKERHUB_TOKEN: ${{ secrets.DOCKERHUB_TOKEN }}
+        shell: bash
+        run: |
+          set -euo pipefail
+          if [ -z "$DOCKERHUB_USERNAME" ]; then
+            echo "No Docker Hub credentials configured; pulling anonymously."
+            exit 0
+          fi
+          printf '%s' "$DOCKERHUB_TOKEN" \
+            | docker login docker.io --username "$DOCKERHUB_USERNAME" --password-stdin
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply .github/workflows/pull-request-validation.yml]: Fixed: the guard now checks both secrets together and errors explicitly if only one is set, instead of proceeding to a login call that would fail on an empty password with no explanation.

File: docs/MERGE_PIPELINE.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  effect, even though it runs.
+- There is no merge queue, so `merge_group` triggers on
+  `pull-request-validation.yml` and `coderabbit-gate.yml` never fire; they
+  exist now so that enabling the queue later does not also require writing
+  new workflow code at the same time, which is the entire point of
+  rehearsing this pipeline here before touching the more important
+  repository.
+- Nothing stops a direct push to `main` today except the pre-commit hook
+  `checklist-git-protected-branches` on a contributor's own machine, which is
+  not a substitute for a server side rule and was never meant to be one.
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply docs/MERGE_PIPELINE.md]: Fixed: hyphenated to "server-side rule".

File: scripts/assert-pin-only-diff.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+# with an alphanumeric and the comparison is line for line either way).
+#
+# `.env.example` is handled separately below rather than through this prefix
+# set, because a bare `VAR=value` prefix would also match GOCRYPTFS_VERSION
+# and DOCKER_IMAGE_TAG_VERSION, and CLAUDE.md and .env.example's own comments
+# both say those two are bumped by hand and are deliberately not tracked by
+# Renovate. Widening this regex to reach them would let a compromised
+# Renovate touch a value it has no business touching and still read as
+# pin-only.
+VERSION = re.compile(r"(?P<prefix>@|\brev:[ \t]+)[0-9A-Za-z][0-9A-Za-z.+_-]*")
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply scripts/assert-pin-only-diff.py]: Fixed: GitHub Actions pins now require the full 40 character commit SHA this repository actually uses, and pre-commit rev: / .tool-versions values now require a real release starting with a digit, so main and similar floating refs no longer normalize as a version bump. Added rejection tests for all three surfaces.

File: scripts/assert-pin-only-diff.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+
+def normalize(line: str, path: str = "") -> str:
+    """Reduce a line to everything about it that a version bump may not change."""
+    if path.endswith(".tool-versions"):
+        return TOOL_VERSION_LINE.sub(r"\g<prefix><version>", line)
+    if path.endswith(".env.example"):
+        var = ENV_VAR_LINE.match(line)
+        if var and var.group("name") in ANNOTATED_ENV_VARS:
+            return re.sub(r'="[^"]*"', '="<version>"', line)
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply scripts/assert-pin-only-diff.py]: Fixed: the annotated env.example value now also has to fully match the release grammar before it is normalized, so a non-version payload in a quoted value leaves the line unchanged and it is refused as a structural mismatch. Added a rejection test.

File: scripts/coderabbit-review-verdict.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+def decide(data: dict) -> tuple[str, str]:
+    """Return (state, description) for `Review Verified`."""
+    if data.get("is_draft"):
+        return "pending", "waiting for ready for review"
+
+    author = data.get("author", "")
+    # Fail closed: an unset or malformed is_fork reads as "is a fork", which
+    # is the direction that cannot mistakenly grant the unattended bot lane.
+    is_fork = bool(data.get("is_fork", True))
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply scripts/coderabbit-review-verdict.py]: Fixed: is_fork now compares against the literal False (`data.get("is_fork", True) is not False`) instead of coercing with bool(), so None, 0, an empty string and similar values read as a fork instead of opening the bot lane. Added a regression test for is_fork: None.

**Comment by ivan-pinatti** [comment]
Addressed all 6 actionable findings from the review (replied on each inline thread) plus the nitpick about `python3` vs `sys.executable` in tests/test_coderabbit_review_verdict.py, which was raised in the review body only, with no separate inline thread to reply on: both test files now use `sys.executable`.

**Comment by ivan-pinatti** [comment]
@coderabbitai review

**Comment by ivan-pinatti** [comment]
@coderabbitai review

**Comment by ivan-pinatti** [comment]
@coderabbitai review
